### ISSUE TITLE
False-positive `type-mismatch` error with generic future

### ISSUE THREAD

**Issue by aidanalphafund** [rust-lang/rust-analyzer#12441]
### Description
It seems rust analyzer can't resolve the output type of a future if that output is generic on another future's output type, leading to  erroneous `type-mismatch` errors.

I initially ran into this using [try_for_each_concurrent](https://docs.rs/futures/latest/futures/prelude/stream/trait.TryStreamExt.html#method.try_for_each_concurrent) on a stream, but managed to narrow the problem to a much simpler reproduction.

### Reproduction ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=203be30ee8419c3fe0c45d09948167a3))
```rust
use std::future::Future;

async fn reproduction() -> usize {
    MyFut(async { 3 }).await
}

struct MyFut<Fut>(Fut);

impl<Fut: Future> Future for MyFut<Fut> {
    type Output = Fut::Output;

    fn poll(
        self: std::pin::Pin<&mut Self>,
        cx: &mut std::task::Context<'_>,
    ) -> std::task::Poll<Self::Output> {
        todo!()
    }
}
```

**rust-analyzer version**: f94fa62d6 2022-05-30 stable
**rustc version**: rustc 1.61.0 (fe5b13d68 2022-05-18)
